### ISSUE TITLE
Training T5 with PPOTrainer: mean_reward stabilizing at the minimum after a few epochs

### ISSUE THREAD

**Issue by PrithwishJana** [huggingface/trl#990]
I am using trl's PPOTrainer to fine-tune a T5 model. Basically, I have a T5 model "`model_t5`" which is pre-trained on a dataset, and an associated RoBERTa tokenizer "`tokenizer`". 

These are my configurations for the PPOTrainer (each sentence in a batch ends with tokenizer.eos_token_id, and is padded with tokenizer.pad_token_id until the max_length of 750. Both batch["input_ids_src"] and batch["input_ids_tgt"] have a max sequence length of 750 tokens.

```
bs = 512
test_dataloader = DataLoader(dataset_train, batch_size = bs) #a dataloader for testing
train_dataloader = DataLoader(dataset_test, batch_size = bs) #a dataloader for training

#----loading my pre-trained tokenizer----
tokenizer = RobertaTokenizer.from_pretrained(tokenizer_path)

#----loading my pre-trained T5 model----
model = T5ForConditionalGeneration.from_pretrained("Salesforce/t5-base")
model.load_state_dict(ckpt)
model_t5 = AutoModelForSeq2SeqLMWithValueHead.from_pretrained(model)
model_t5_ref = create_reference_model(model_t5)

#----initializing a PPO config----
PPO_config = PPOConfig(
    steps=20000, learning_rate=1.41e-5, remove_unused_columns=False, batch_size = bs, log_with="wandb")

#----initializing a PPO trainer----
PPO_trainer = PPOTrainer(config = PPO_config, 
                            model = model_t5, 
                            ref_model = model_t5_ref, 
                            tokenizer = tokenizer
                        )

#----enabling training with accelerator----
if PPO_trainer.accelerator.num_processes == 1:
    device = 0 if torch.cuda.is_available() else "cpu"
else:
    device = PPO_trainer.accelerator.device

#----preparing train_dataloader with the accelerator----
train_dataloader = PPO_trainer.accelerator.prepare(train_dataloader)

#----generation_kwargs as per suggestions from https://huggingface.co/docs/trl/main/en/how_to_train----
#----only changed pad_token_id to tokenizer.pad_token_id, instead of tokenizer.eos_token_id
#----and max_new_tokens to 750, from 32
generation_kwargs = {
    "min_length": -1, # don't ignore the EOS token
    "top_k": 0.0, # no top-k sampling
    "top_p": 1.0, # no nucleus sampling
    "do_sample": True, # yes, we want to sample
    "pad_token_id": tokenizer.pad_token_id,
    "max_new_tokens": 750
}

#----generation_kwargs for testing, with do_sample as False----
test_kwargs = {
    "do_sample": False,
    "max_new_tokens": 750
}
```

The training loop goes something like this: (basically, I have a calculateReward() function that returns the rewards as a list of tensors, each being a floating-point value between -0.5 and 1.5.) (Also, I didn't create "query" and "response" as keys in my batch, hope that's not an issue)

```
for epoch in range(num_epochs):
    for batch in tqdm(train_dataloader):
        src_id_list = list(batch["input_ids_src"])
        tgt_id_list = list(batch["input_ids_tgt"])

        tgtPred_id_list = PPO_trainer.generate(src_id_list, **generation_kwargs) 

        rewards = calculateReward(tgtPred_id_list, tgt_id_list)

        stats = PPO_trainer.step(src_id_list, tgtPred_id_list, rewards)
        PPO_trainer.log_stats(stats, batch, rewards)

    test_func(test_dataloader, model_t5, test_kwargs, tokenizer)
```

But when I train this model, surprisingly the mean reward settles at -0.5 (the lowest reward) after a few steps of fluctuation. I am not sure why this is happening, I believe to follow all the guidelines from the TRL documentation and FAQ.
<img width="1233" alt="Screenshot 2023-11-14 at 5 57 53 AM" src="https://github.com/huggingface/trl/assets/26099047/f53e35e0-133c-498f-86be-9ebd2a82ae11">

I also went through the per-epoch test output. After 10 steps, the outputs were okayish with some sentences ending with lots of repeated words. But after 20/30 steps, all test sentences were either empty or at most of length 1 or 2.

Is it because I didn't set the attention_mask during training, and thus the model is now prioritizing the padding_tokens? For trl, do we need to supply the attention_mask to mask the padding tokens, or is it calculated automatically? 

Please suggest what could have gone wrong in this pipeline, or any insights or suggestions on potential sources of error or areas to investigate. I'm struggling with my current pipeline, and trying to figure out what's gone wrong. Any kind of help would be immensely appreciated. Thanks!

**Comment by imgremlin** [comment]
I have the same issue. I'm using BartForConditionalGeneration, and for the last 3 weeks trying to debug why my rewards drop just after the first step. Have you solved the issue?
![image](https://github.com/huggingface/trl/assets/40468037/94b32492-751c-4051-aa76-050cfce1e213)

**Comment by MisterPANDC** [comment]
I also met the same issue using PPOTrainer to train seq2seq models, and I also found something wrong with padding and attention_mask like @PrithwishJana said.  Is this issue caused by paddings? Have you solved this problem?
